### PR TITLE
Use cublasSgemmEx when input dtype is float16

### PR DESCRIPTION
**Pull request by okuta**
Cupy use cublasSgemmEx to reduce memory usage.

### CODE REVIEW

File: cupy/core/core.pyx [hunk 1]
```diff
@@ -1933,89 +1942,23 @@ cpdef ndarray tensordot_core(
         c.shape = (n, m)
 
     # Be careful that cuBLAS uses the FORTRAN-order matrix representation.
-    if k == 1:
```
Comment (unnonouno) [inline cupy/core/core.pyx]: Why is this code removed?

Comment (okuta) [inline cupy/core/core.pyx]: It is historical reasons.
We had believed that `semm` can not calculate between a matrix and a vector.

**Comment by gwtnb** [comment]
`flake8 --config=.flake8.cython`

File: cupy/cuda/cupy_cuda.h [hunk 2]
```diff
@@ -406,6 +430,13 @@ int cublasSgemmBatched(
     return 0;
 }
 
+int cublasSgemmEx(
+        cublasHandle_t handle, cublasOperation_t transa,
+        cublasOperation_t transb, int m, int n, int k,
+        const float *alpha, const void *A, cudaDataType Atype,
+        int lda, const void *B, cudaDataType Btype, int ldb,
+        const float *beta, void *C, cudaDataType Ctype, int ldc);
+
```
Comment (gwtnb) [inline cupy/cuda/cupy_cuda.h]: `cublassSgemmEx` is a BLAS-like extension, so this declaration should be moved.

File: cupy/core/core.pyx [hunk 3]
```diff
@@ -1956,32 +1957,45 @@ cpdef ndarray dot(ndarray a, ndarray b, ndarray out=None):
...
 
+    global _cuda_runtime_version
+    if _cuda_runtime_version is None:
+        _cuda_runtime_version = runtime.runtimeGetVersion()
+
+    use_sgemmEx = (_cuda_runtime_version >= 7500 and
+                   a.dtype == 'e' and b.dtype == 'e' and
+                   (ret_dtype == 'e' or ret_dtype == 'f'))
+
+    # Cast to float32 or float64
```
Comment (gwtnb) [inline cupy/core/core.pyx]: This comment should be updated.

**Comment by gwtnb** [comment]
LGTM except the above comment.

**Pull request merged**